File: backend/app/api/graph.py
"""`/graph` router — the Cosmograph projection (M6).

Projects Neo4j into `{nodes, links}` for `@cosmograph/react`. Kept legible by default: top entities
(by interaction degree) + the events that mention them, capped to `limit`, optionally filtered by a
recency `window_days` and a `priority` class. Degrades to an empty, `available: false` graph when
Neo4j is unreachable.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from app.api.deps import graph_store_dep
from app.graph import GraphStore
from app.schemas.api import GraphLinkOut, GraphNodeOut, GraphOut

router = APIRouter(tags=["graph"])


@router.get("/graph", response_model=GraphOut)
def get_graph(
    limit: int = Query(default=150, ge=1, le=1000, description="Max nodes (kept readable)."),
    window_days: int | None = Query(default=None, description="Only edges newer than N days."),
    priority: str | None = Query(default=None, description="Only events of this priority class."),
    store: GraphStore | None = Depends(graph_store_dep),
) -> GraphOut:
    """Return capped nodes/links for the network + timeline view."""
    if store is None:
        return GraphOut(nodes=[], links=[], truncated=False, available=False)
    view = store.graph_view(limit=limit, window_days=window_days, priority=priority)
    return GraphOut(
        nodes=[GraphNodeOut(**n.__dict__) for n in view.nodes],
        links=[GraphLinkOut(**l.__dict__) for l in view.links],
        truncated=view.truncated,
        available=True,
    )

File: frontend/src/api/graph.ts
import { z } from "zod";
import { apiGet } from "./client";

/**
 * GET /graph (M6) — the Cosmograph projection: capped nodes/links from Neo4j (events + top
 * entities). `available` is false when Neo4j is unreachable (degraded, empty graph).
 */
export const GraphNodeSchema = z.object({
  id: z.string(),
  label: z.string(),
  kind: z.string(),
  type: z.string().nullable().optional(),
  priority_class: z.string().nullable().optional(),
  ts: z.string().nullable().optional(),
});

export const GraphLinkSchema = z.object({
  source: z.string(),
  target: z.string(),
  kind: z.string(),
  ts: z.string().nullable().optional(),
});

export const GraphSchema = z.object({
  nodes: z.array(GraphNodeSchema),
  links: z.array(GraphLinkSchema),
  truncated: z.boolean(),
  available: z.boolean(),
});

export type GraphNode = z.infer<typeof GraphNodeSchema>;
export type GraphLink = z.infer<typeof GraphLinkSchema>;
export type GraphData = z.infer<typeof GraphSchema>;

export interface GraphQuery {
  limit?: number;
  windowDays?: number;
  priority?: string;
}

export function fetchGraph(query: GraphQuery = {}): Promise<GraphData> {
  const params = new URLSearchParams();
  if (query.limit) params.set("limit", String(query.limit));
  if (query.windowDays) params.set("window_days", String(query.windowDays));
  if (query.priority) params.set("priority", query.priority);
  const qs = params.toString();
  return apiGet(`/graph${qs ? `?${qs}` : ""}`, GraphSchema);
}
